### PR TITLE
Overriding user-agent when using cy.request or cy.visit

### LINKED ISSUE #3719: visit() does not send user agent header as specified
### Current behavior:

With the new feature in 3.2.0 that enables setting headers for `visit()`, "user agent" header is not correctly passed to server side 

### Desired behavior:

user agent header can be passed to server side as specified 

### Steps to reproduce: (app code and test code)
Using the following code: 
```js
cy.visit({
      url: '/',
      headers: {
        'User-Agent': 'bad browser',
        'hello': 'hi'
      }
    })
```
And observe server side receives something like this:
image

As shown above, header "hello" is correctly passed, therefore the header setting is recognized by cypress. However, header "User-Agent" is still my real browser which is the error  

### Versions

cypress 3.2.0

### LINKED ISSUE #3873: Overriding user-agent when using cy.request
### Current behavior:

`cy.request` overrides the **user-agent** if it is provided in the `options.headers` object of the request

### Desired behavior:

If the `User-agent` is explicitly set in `options.headers` object when using `cy.request` it should not be overwritten with the current browsing session's **user-agent**.

### Steps to reproduce: (app code and test code)

e.g:

This implementation
```js
cy.request({
  url,
  method: 'GET',
  headers: {
    'user-agent': 'test-agent',
  },
}).then(response => {
  expect(response.requestHeaders).to.have.property('user-agent', 'test-agent');
});
```

leads to:
Screen Shot 2019-04-03 at 17 10 44

### Versions

Cypress: 3.1.5
OS: MacOS 10.13.3
Browser: Chrome - Version 73.0.3683.86

### PR DESCRIPTION
**Pull request by AbdullahZN**
Added possibility to override **user-agent** headers for `request.send`annd `request.sendStream`

Since headers can be provided by the client either in lowercase/uppercase/mix of both:
- I also added a method to lowercase all the provided headers

💡 I don't know where this kind of utils method should go, is it fine in the `request.coffee` ?

Closes #3873 
Closes #3719

### CODE REVIEW

**Comment by jennifer-shehane** [comment]
Other issues regarding lowercase request headers:
- https://github.com/cypress-io/cypress/issues/3637
- https://github.com/cypress-io/cypress/issues/2879

**Comment by AbdullahZN** [comment]
Hi, 
Any update on this ?

File: packages/server/lib/request.coffee [hunk 1]
```diff
@@ -130,6 +130,12 @@ reduceCookieToArray = (c) ->
 createCookieString = (c) ->
   reduceCookieToArray(c).join("; ")
 
+propertiesToLowerCase = (obj) ->
```
Comment (flotwig) [inline packages/server/lib/request.coffee]: This mutates `options.headers`. We probably don't want to change the casing of every key in the object in light of #2879.

Perhaps just making a case-insensitive getter function would be better, something like `caseInsensitiveGet`?

File: packages/server/lib/request.coffee [hunk 2]
```diff
@@ -271,7 +277,8 @@ module.exports = (options = {}) ->
         jar: true
       }
 
-      if ua = headers["user-agent"]
+      propertiesToLowerCase(options.headers)
```
Comment (flotwig) [inline packages/server/lib/request.coffee]:
```coffeescript
if not caseInsensitiveGet("user-agent") and...
```

File: packages/server/lib/request.coffee [hunk 3]
```diff
@@ -332,7 +339,8 @@ module.exports = (options = {}) ->
         followRedirect: true
       }
 
-      if ua = headers["user-agent"]
+      propertiesToLowerCase(options.headers)
```
Comment (flotwig) [inline packages/server/lib/request.coffee]: same as before

**Comment by AbdullahZN** [comment]
Thanks for the review @flotwig, corrected 👍

**Comment by flotwig** [comment]
Looks good @AbdullahZN, but the same unit test is failing every time this PR runs, so that will need to be figured out before this can be merged.

**Comment by AbdullahZN** [comment]
I'll dig into that. It's hard to debug though: tests are failing on my local develop branch 🤔

**Comment by AbdullahZN** [comment]
@flotwig It's ok now, tests are green 🍏 :)

File: packages/server/test/unit/request_spec.coffee [hunk 4]
```diff
@@ -315,6 +315,43 @@ describe "lib/request", ->
       .then (resp) ->
         expect(resp.body).to.eq("it worked")
 
+    it "lower cases headers", ->
+      nock("http://localhost:8080")
+      .matchHeader("test", "true")
+      .get("/foo")
+      .reply(200, "derp")
+
+      headers = {}
```
Comment (flotwig) [inline packages/server/test/unit/request_spec.coffee]: `headers` doesn't look like it's used anywhere

**Comment by flotwig** [comment]
I think this will also fix #3719, since cy.visit() goes through request. Unless there's some other place where user-agent gets written.

**Comment by AbdullahZN** [comment]
It will require a bit more work. I'll try to handle it in another PR

**Comment by flotwig** [comment]
@AbdullahZN I added a test in your branch for cy.visit user-agent, it appears to be passing with just the changes you've made: https://github.com/cypress-io/cypress/pull/3920/commits/3d1c8320446114a487387bd690cd16acc9840661

We can close #3719 too then, nice

**Comment by flotwig** [comment]
> Since headers can be provided by the client either in lowercase/uppercase/mix of both:
- I also added a method to lowercase all the provided headers

@AbdullahZN I just learned that Express's `req.get()` is case-insensitive, so we could've used that all along :smile:

**Pull request merged**